### PR TITLE
MDEV-17038 ALTER TABLE CHANGE COLUMN c1 c1 bigint NOT NULL - generates error if table uses SYSTEM VERSIONING

### PR DESCRIPTION
**Pull request by midenok**
@dr-m please review.

This code is submitted under the BSD-new license.

### CODE REVIEW

**Comment by dr-m** [comment]
Regarding code formatting, the InnoDB rules for lines split after unbalanced `(` are a bit convoluted.
If there is an unbalanced `(` in a line, then subsequent lines should be aligned with that parenthesis, indenting the line first with TABs (8 columns each), and then with spaces as necessary.
But, there is the exception that if the line is terminated in an unbalanced `(`, we will indent with extra 8 columns instead.

Around 2005, everyone agreed to use these GNU Emacs settings, which implement the above:
```elisp
(require 'cc-mode)

(defun innodb-cc-mode (&optional map)
  "InnoDB adjustments to cc-mode.  In MAP, binds C-c C-c to 'compile."
  (if (and
       (buffer-file-name)
       (string-match "/\\(innobase\\|xtradb\\).*/.*[ihc]\\'"
		     (buffer-file-name)))
      (progn
	(setq
	 mode-name (concat "Ib" mode-name)
	 c-basic-offset 8
	 c-label-minimum-indentation 0)
	(add-to-list 'c-offsets-alist '(c . 0))
	(add-to-list 'c-offsets-alist '(arglist-intro . +))
	(add-to-list 'c-offsets-alist '(label . [0]))))
  (if map
      (define-key map "\C-c\C-c" 'compile)))

(add-hook 'c-mode-hook (function (lambda () (innodb-cc-mode c-mode-map))))
(add-hook 'c++-mode-hook (function (lambda () (innodb-cc-mode c++-mode-map))))

(add-to-list 'auto-mode-alist '("\\.ic\\'" . c++-mode))
(add-to-list 'auto-mode-alist '("storage/innobase/.*\\.h\\'" . c++-mode))
```

File: storage/innobase/handler/handler0alter.cc [hunk 1]
```diff
@@ -8959,7 +8966,8 @@ commit_try_norebuild(
 		DBUG_RETURN(true);
 	}
 
-	if (change_fields_versioning_try(ha_alter_info, ctx, trx, old_table)) {
+	if ((ha_alter_info->handler_flags & ALTER_COLUMN_UNVERSIONED) &&
+		vers_change_fields(ha_alter_info, ctx, trx, old_table)) {
 		DBUG_RETURN(true);
```
[the remark this answers is not in the document]
Comment (dr-m) [reply storage/innobase/handler/handler0alter.cc]: The 8-column indentation applies to lines ending in an open parenthesis, bracket or brace.
I posted an explanation of the InnoDB code formatting (along with a snippet from my `$HOME/.emacs`) at the top level.

**Comment by midenok** [comment]
> If there is an unbalanced ( in a line, then subsequent lines should be aligned with that parenthesis, ...

Aligned with next char after `(`. This indentation style is mostly used by GNU I believe, a bit unpractical by fact. Yes, I reindented the patch like that.

File: storage/innobase/handler/handler0alter.cc [hunk 2]
```diff
@@ -866,11 +871,10 @@ ha_innobase::check_if_supported_inplace_alter(
 
 	const bool need_rebuild = innobase_need_rebuild(ha_alter_info, table);
 
-	if (need_rebuild
-	    && (table->versioned(VERS_TIMESTAMP)
-		|| altered_table->versioned(VERS_TIMESTAMP))) {
+	if (altered_table->versioned(VERS_TIMESTAMP)
+	    && (ha_alter_info->handler_flags & INNOBASE_ALTER_VERSIONED_REBUILD)) {
```
Comment (dr-m) [inline storage/innobase/handler/handler0alter.cc]: I would keep the `need_rebuild &&` condition. I see no reason to block (for example) changes of column comments or default values. Tests passed for me when I restored that condition.

**Comment by dr-m** [comment]
Squashed into commit 4acfc6ecd9085ff3307aa741351924287fb9593c.

**Comment by laurynas-biveinis** [comment]
@dr-m is there any public source for the formatting settings, like the Emacs config above? Thank you

**Pull request closed without merging**